### PR TITLE
Make sure IApplicationLifetime gets added even if startup throws.

### PR DESCRIPTION
**Pull request by Tratcher**
#434 @davidfowl @HaoK

### CODE REVIEW

**Comment by davidfowl** [comment]
Add a test

**Comment by Tratcher** [comment]
Updated

File: test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs [hunk 1]
```diff
@@ -74,6 +75,20 @@ public void Build_honors_UseStartup_with_string()
         }
 
         [Fact]
+        public async Task IApplicationLifetimeRegisteredEvenWhenStartupCtorThrows_Fallback()
+        {
+            var builder = CreateWebHostBuilder();
+            var serverFactory = new TestServerFactory();
+            var engine = (HostingEngine)builder.UseServer(serverFactory).UseStartup<StartupCtorThrows>().Build();
```
Comment (davidfowl) [inline test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs]: Why does this need to cast?

File: src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs [hunk 2]
```diff
@@ -70,6 +70,7 @@ public class HostingEngine : IHostingEngine
             _startupLoader = startupLoader;
             _captureStartupErrors = captureStartupErrors;
             _applicationLifetime = new ApplicationLifetime();
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs]: Is there any reason not to add this in the hosting services?

Comment (Tratcher) [inline src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs]: Yes, it needs the actual instance to call things like NotifyStarted.

File: test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs (hunk 1 above)
Comment (Tratcher) [inline test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs]: Some of the tests assert on non-interface members, but not this one. removed.

**Comment by Tratcher** [comment]
Updated.

**Comment by davidfowl** [comment]
Why is travis failing?

**Comment by Tratcher** [comment]
IHttpContextFactory due to the lack of a recent coherence build.

**Comment by Tratcher** [comment]
@muratg

**Comment by HaoK** [comment]
:shipit:

**Pull request merged**